replica/database.cc

```diff
@@ -1422,7 +1422,18 @@ future<> database::create_in_memory_keyspace(const lw_shared_ptr<keyspace_metada
 }
 
 future<> database::modify_keyspace_on_all_shards(sharded<database>& sharded_db, std::function<future<>(replica::database&)> func, std::function<future<>(replica::database&)> notifier) {
-    co_await sharded_db.invoke_on_all(func);
+    // Run func first on shard 0
+    // to allow "seeding" of the effective_replication_map
+    // with a new e_r_m instance.
+    co_await sharded_db.invoke_on(0, func);
```

Comment:
> > It is more efficient to do that first on shard 0
> > and then on all other shards, otherwise multiple
> > shards might need to calculate to new e_r_m (and reach
> > the same result).  When the new e_r_m is "seeded" on
> > shard 0, other shards will find it there and clone
> > a local copy of it - which is ore efficient.
> 
> Why can't we rely on listeners to do that?

To do what?  invoke_on shard 0?
We can, but that will just replicate the logic in a lower level.
since `modify_keyspace_on_all_shards` is defined as private I think it makes sense to centralize the logic in a middle layer function.

> keyspace::create_replication_strategy() already has a preemption point between _replication_strategy is assigned and update_effective_replication_map(). So users may observe inconsistent state. They could just as well wait a bit longer until all shards have updated metadata. Or is that an unintended bug?

It is unintended.

Reply:
kept `modify_keyspace_on_all_shards` in v4

Reply:
> To do what?  invoke_on shard 0?

I meant to replicate erm. The listener could replicate erm, in which case we can ditch _erm_factory and not special-case shard 0 here. But if we need to change _replication_strategy atomically, that won't do.